### PR TITLE
CoW: Track references in unstack if there is no copy

### PR DESCRIPTION
**Pull request by phofl**
- [ ] xref #56633
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jorisvandenbossche @rhshadrach 

@rhshadrach was correct, we can get there without making a copy. it's a special case, but nevertheless

### CODE REVIEW

**Comment by phofl** [comment]
might be worth back porting, but not necessary

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -229,20 +230,33 @@ def arange_result(self) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.bool_]]:
...
-            values, index=index, columns=columns, dtype=values.dtype
+        result = self.constructor(
+            new_values, index=index, columns=columns, dtype=new_values.dtype, copy=False
         )
+        if isinstance(values, np.ndarray):
+            base, new_base = values.base, new_values.base
+        elif isinstance(values, NDArrayBackedExtensionArray):
+            base, new_base = values._ndarray.base, new_values._ndarray.base
+        else:
+            base, new_base = 1, 2  # type: ignore[assignment]
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/reshape.py:253]: Are we sure this base checking always works? (compared to doing something like `np.shares_memory`) 
From some quick tests, it seems that even if you chain multiple no-copy transformations (like we do with `reshape().swapaxes().reshape()`), the base object stays the same.

If the base object is backed by something non-numpy (eg parrow), this doesn't seem to be the case, but, I think in a DataFrame we probably always end up with an array with a base object that is another numpy array.

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -229,20 +230,33 @@ def arange_result(self) -> tuple[npt.NDArray[np.intp], npt.NDArray[np.bool_]]:
...
+        if base is new_base:
+            # We can only get here if one of the dimensions is size 1
+            mgr = result._mgr
+            mgr.blocks[0].refs = obj._mgr.blocks[0].refs
+            mgr.blocks[0].refs.add_reference(mgr.blocks[0])
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/reshape.py:258]: Can this use `result._mgr.add_references(obj._mgr)` ?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for looking into this!

File: pandas/core/reshape/reshape.py (hunk 1 above)
Comment (phofl) [reply pandas/core/reshape/reshape.py:253]: we reshape into a 3dim array in between, so anything that is 1d copies there

File: pandas/core/reshape/reshape.py (hunk 2 above)
Comment (phofl) [reply pandas/core/reshape/reshape.py]: oh yeah thanks, I was looking for this one but couldn't find it (was looking in the wrong place..)

**Comment by phofl** [comment]
this one should be good to merge

**Review by mroeschke — APPROVED** [review]
LGTM

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**